### PR TITLE
Reduce TIME-WAIT load

### PR DESCRIPTION
**Pull request by rzezeski**
This addresses #367 

A combination of ibrowse's inefficient load balancing algorithm and
default socket inactivity timeout of 10 seconds can cause TIME-WAIT
load.  It becomes worse as the pool size of pipeline sizes are
increased.  This patch is a temporary workaround to reduce socket
churn and thus TIME-WAIT load.  It does so by increasing the
inactivity timeout to 600 seconds across the board.  Below is a chart
of showing the amount of socket churn in connections per minute for
the different timeout values both at idle and while under load from
basho bench.  These numbers were calculated by a DTrace script which
counted the number of new connections being accepted on port 8093.

| Timeout | Socket Churn At Idle | Socket Churn Under Load |
| --- | --- | --- |
| 10s | ~59 conns/min | ~29 conns/min |
| 600s | ~9 conns/min | ~6 conns/min |

The timeout is set via application env because ibrowse has the
absolute most complex configuration management code I have ever seen
and this was the easiest way to make sure the timeout is set
correctly.  This is just a workaround until after 2.0 when other HTTP
clients and pools may be tested.  ibrowse seems to have many issues,
this is but just one.

For more background see the following issues:

https://github.com/basho/yokozuna/issues/367
https://github.com/basho/yokozuna/pull/358
https://github.com/basho/yokozuna/issues/330
https://github.com/basho/yokozuna/issues/320

### CODE REVIEW

**Comment by rzezeski** [comment]
## 600s timeout

Idle:

```
root@oldboy:~# dtrace -s /home/ryan/tcp-conn-rate.d
Monitor rate of ESTABLISHED passive connections over 1 minute............................................................

127.0.0.1:8093 9
```

Load:

```
root@oldboy:~# dtrace -s /home/ryan/tcp-conn-rate.d
Monitor rate of ESTABLISHED passive connections over 1 minute............................................................

127.0.0.1:8093 6
```

## 10s timeout

Idle:

```
root@oldboy:~# dtrace -s /home/ryan/tcp-conn-rate.d
Monitor rate of ESTABLISHED passive connections over 1 minute............................................................

127.0.0.1:8093 59
```

Load:

```
root@oldboy:~# dtrace -s /home/ryan/tcp-conn-rate.d
Monitor rate of ESTABLISHED passive connections over 1 minute............................................................

127.0.0.1:8093 29
```

**Comment by rzezeski** [comment]
@wbrown If you get a chance you might want to give this PR a spin.  It is not a perfect patch but it will have to be enough until after 2.0 is out and we can look at replacing the entire HTTP client.

File: src/yz_app.erl [hunk 1]
```diff
@@ -38,6 +38,7 @@ start(_StartType, _StartArgs) ->
     Enabled = ?YZ_ENABLED,
     case yz_sup:start_link(Enabled) of
         {ok, Pid} ->
+            _ = application:set_env(ibrowse, inactivity_timeout, 600000),
```
Comment (coderoshi) [inline src/yz_app.erl:41]: This is 600 secs, since I assume this is ms

Comment (rzezeski) [inline src/yz_app.erl:41]: (facepalm) This is what I get for leaving a branch for over a week.  Yea, so 600s is fine too. I actually wanted to set it to infinity but ibrowse is has a bug where if you set the value via app env it won't accept `infinity` (but it does if you set it other ways, but those other ways only work for some connections...ibrowse literally has the most asinine code for its configuration I want to break shit every time I think about it).

**Comment by rzezeski** [comment]
I amended the commit message to say 600 seconds.

**Comment by coderoshi** [comment]
@rzezeski could you post `tcp-conn-rate.d` in a gist? I'd like to test this.

**Comment by rzezeski** [comment]
> @rzezeski could you post tcp-conn-rate.d in a gist? I'd like to test this.

See the end of this comment: https://github.com/basho/yokozuna/issues/330#issuecomment-40783996

**Comment by coderoshi** [comment]
+1 b02c0bd

I saw a similar pattern running this dtrace.

**600s timeout Idle**

```
127.0.0.1:10014 9
```

**600s timeout Load**

```
127.0.0.1:10014 9
```

**10s timeout Idle**

```
127.0.0.1:10014 58
```

**10s timeout Load**

```
127.0.0.1:10014 55
```

I had to make a small change to the script to work in OSX (`tcp:::accept-established` doesn't exist in OSX)

```
#pragma D option quiet

dtrace:::BEGIN
{
    printf("Monitor rate of ESTABLISHED passive connections over 1 minute");
}

tcp:::state-change
/args[2]->tcps_state == "state-established"/
{
    @est[args[1]->ip_saddr, args[1]->sport] = count();
}

profile:::tick-1sec
{
    printf(".");
}

profile:::tick-60sec
{
    printf("\n\n");
    printa("%s:%d %@d\n", @est);
    exit(0);
}
```

**Comment by rzezeski** [comment]
@borshop merge

**Pull request merged**